cmd/occollector/app/collector/collector.go

```diff
@@ -161,9 +181,19 @@ func execute() {
 	receiversCloseFns := createReceivers(spanProcessor)
 	closeFns = append(closeFns, receiversCloseFns...)
 
-	logger.Info("Collector is up and running.")
+	err = initTelemetry(telemetryLevel, v.GetInt(metricsPortCfg), asyncErrorChannel, logger)
+	if err != nil {
+		logger.Error("Failed to initialize telemetry", zap.Error(err))
+		os.Exit(1)
+	}
+
+	select {
+	case err = <-asyncErrorChannel:
+		logger.Error("Asynchronous error received, terminating process", zap.Error(err))
+	case <-signalsChannel:
+		logger.Info("Received kill signal from OS")
+	}
 
-	<-signalsChannel
 	logger.Info("Starting shutdown...")
 
 	// TODO: orderly shutdown: first receivers, then flushing pipelines giving
```

Comment:
at this point, shouldn't you also gracefully shutdown the http server that is serving metrics?

Reply:
(shutdown would ideally include waiting the http server to exit before exiting the process)

Reply:
To gracefully shutdown here we will need to start having a timeout and so on, just calling Close is a rude shutdown (ie.: underlying listener is closed). This is desirable but we didn't do it yet for the pipes sending data to the backend. I think that adding it now for the handler of Prometheus metrics is an overkill (right now the worst that can happen is a GET to be cut in the middle and Prometheus losing one last scrape of data, but that was already in a race).